_ap/18/index.adoc

```diff
@@ -3,76 +3,200 @@ num: 18
-
-## Participants
-
-// What components and/or teams are directly involved
-//
-// For this metrics example this would include the customer and the control plane via which
-// the metrics are exposed
-
-## Trade-offs
-
-// Explain any consequences of using this pattern. 
-// Be sure to cover both pros and cons:
-// * what this pattern enables and the benefits it brings
-// * what this pattern prevents or makes harder
-//
-// For this metrics example such benefits would include:
-// * A mechanism for API evolution
-// * Consistency in naming of metrics and labels within this service
-// * Consistency with other services which also apply this AP
-//
-// And on the down side:
-// * The Prometheus format eventually be superseded
-// * Naming and consistency is hard, so multiple versions of the API 
-//   might be needed to achieve those goals. 
+== Context and Problem Statement
+
+We are internally committed to using OpenAPI specification and encourage our customers to do the same with products like API Designer and Service Registry.
+
+The current approach to automatically generating client SDKs for Red Hat's application services has a number of problems/issues that need to be holistically addressed.
+This AP will document the challenges with our current approach and propose a solution using new technologies.
+
+In client-server architectures, servers are written using either a contract-first or code-first approach.
+However, we always have the need to interact with those APIs from clients.
+Generation of the API clients can (and should) be completely automated.
+This represents a contract-first approach to creating API clients.
+
+We do generate/write clients for several purposes, the most relevant are:
+
+* Service SDKs
+* CLI tools
+* UI (e.g. console.dot)
+* Provisioning Tools (Ansible, Terraform)
+
+Most of our toolchains are (in some ways) using `openapi-generator`, developed and maintained by the "OpenAPI Tools" organization (which is NOT affiliated with the OpenAPI initiative).
+
+Unfortunately, OpenAPI Generator is not a well maintained OSS project:
+image::openapi-generator-gh.png[]
+
+Even our own attempts to fix known bugs have been ineffective:
+https://github.com/OpenAPITools/openapi-generator/pull/13616
+
+It's also pretty common to have breaking changes in the generator tooling, which results in breaking changes in the generated code.
+When this happens it can be hard to detect and often happens without a clear migration path.
+
+For these reasons we are often https://github.com/redhat-developer/app-services-sdk-go/blob/13d1cad9a937c456a9623ad188c44790e3370d83/openapitools.json#L5[stuck on old versions], preventing (in some cases) the opportunity to fix CVEs and issues by keeping transitive dependencies and used frameworks updated.
+
+Following are a variety of critical/important issues we currently have due to our OpenAPI Generator based approach to API Client generation.
+
+=== Security
+
```

Comment:
Discuss mitigation on security issues coming from. Probably iso risk between old and new approach but make it clear of the security risk paths.

Reply:
Ah it's in the Thread Model section